src/js/actions/libraries.js

```diff
@@ -499,6 +497,145 @@ define(function (require, exports) {
+     * in Photoshop. The library store will listen to the document save and close event. 
+     * 
+     * @param {AdobeLibraryElement} element
+     * @return {Promise}
+     */
+    var openGraphicForEdit = function (element) {
+        var representation = element.getPrimaryRepresentation();
+        
+        if (representation.type !== _REP_PHOTOSHOP_TYPE) {
+            // Due to the limitation in DS, we only support editing psd file.
+            log.debug("[CC Lib] unsupported graphic type: " + representation.type);
+            return Promise.resolve();
+        }
+        
+        var tempFilePath,
+            tempPreviewPath;
+        
+        return Promise
+            .bind(this)
+            .then(function () {
+                var graphicEditStatus = this.flux.stores.library.getGraphicEditStatus(element);
+                
+                // If the element already have a temp file, update the temp file path.
+                if (graphicEditStatus) {
+                    tempFilePath = graphicEditStatus.documentPath;
+                    tempPreviewPath = graphicEditStatus.previewPath;
+                    return;
+                }
+                
+                // Otherwise, create a temp copy of the element for edit.
+                return _getTempPaths()
+                    .bind(this)
+                    .then(function (paths) {
+                        var tempFilename = this.flux.stores.library.generateTempFilename(element);
+                        
+                        tempFilePath = paths.tempBasePath + pathUtil.sep + tempFilename;
+                        tempPreviewPath = tempFilePath.replace(".psd", "p.png");
+                    })
+                    .then(function () {
+                        return Promise.fromNode(function (done) {
+                            element.getPrimaryRepresentation().getContentPath(done);
+                        }).then(function (contentPath) {
+                            if (!contentPath) {
+                                return Promise.reject("Failed to fetch content of element: " + element.name);
+                            }
+                            
+                            return contentPath;
+                        });
+                    })
+                    .then(function (contentPath) {
+                        return this.flux.actions.export.copyFile(contentPath, tempFilePath);
+                    });
+            })
+            .then(function () {
+                var previewSetting = {
+                    path: tempPreviewPath,
+                    width: RENDITION_GRAPHIC_SIZE,
+                    height: RENDITION_GRAPHIC_SIZE
+                };
+                
+                // Open the graphic asset's temp file and tell PS to generate an updated 
+                // preview whenever the file is saved
+                return this.flux.actions.documents.open(tempFilePath, {
```

Comment:
You want to transfer here, enqueueing one action from another is a good way to get deadlocks.